File: internal/core/domain/transport.go
package domain

// TransportState is the modeled (open-loop) play state of the deck.
type TransportState int

const (
	StateStopped TransportState = iota
	StatePlaying
)

// HyperDeckStatus maps the modeled state to a HyperDeck transport "status" value.
func (s TransportState) HyperDeckStatus() string {
	if s == StatePlaying {
		return "play"
	}
	return "stopped"
}

// TransportInfo is the payload of a "transport info" response.
type TransportInfo struct {
	Status string
	Speed  int
	ClipID int
	SlotID int
}

// SlotInfo is the payload of a "slot info" response.
type SlotInfo struct {
	Present bool
	SlotID  int
}

// DeviceInfo is the payload of a "device info" response.
type DeviceInfo struct {
	ProtocolVersion string
	Model           string
	UniqueID        string
}

File: crates/hyperdeck-core/src/domain/transport.rs
//! Transport state and response payloads (port of `internal/core/domain/transport.go`).

/// The modeled (open-loop) play state of the deck.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum TransportState {
    #[default]
    Stopped,
    Playing,
}

impl TransportState {
    /// Maps the modeled state to a HyperDeck transport "status" value.
    pub fn hyperdeck_status(self) -> &'static str {
        match self {
            TransportState::Playing => "play",
            TransportState::Stopped => "stopped",
        }
    }
}

/// The payload of a `transport info` response.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct TransportInfo {
    pub status: String,
    pub speed: i32,
    pub clip_id: i32,
    pub slot_id: i32,
}

/// The payload of a `slot info` response.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct SlotInfo {
    pub present: bool,
    pub slot_id: i32,
}

/// The payload of a `device info` response.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct DeviceInfo {
    pub protocol_version: String,
    pub model: String,
    pub unique_id: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hyperdeck_status_strings() {
        assert_eq!(TransportState::Stopped.hyperdeck_status(), "stopped");
        assert_eq!(TransportState::Playing.hyperdeck_status(), "play");
    }
}
